crypto/pkcs7/bio/cipher.c

```diff
@@ -0,0 +1,310 @@
+  if (ctx == NULL || ctx->cipher == NULL || ctx->done || !ctx->ok) {
+    return 0;
+  }
+  BIO *next = BIO_next(b);
+  if (next == NULL) {
+    return 0;
+  }
+
+  int bytes_consumed = 0;
+  int remaining = inl;
+  const int max_crypt_size =
+      (int)sizeof(ctx->buf) - EVP_CIPHER_CTX_block_size(ctx->cipher) + 1;
+  while (remaining > 0) {
+    assert(bytes_consumed + remaining == inl);
+    if (ctx->buf_len == 0) {
+      ctx->buf_off = 0;
+      int to_encrypt = remaining < max_crypt_size ? remaining : max_crypt_size;
+      uint8_t *in_pos = ((uint8_t *)in) + bytes_consumed;
+      ctx->ok = EVP_EncryptUpdate(ctx->cipher, ctx->buf, &ctx->buf_len, in_pos,
+                                  to_encrypt);
+      if (!ctx->ok) {
+        break;
+      };
+      bytes_consumed += to_encrypt;
+      remaining -= to_encrypt;
+    }
+    int bytes_written = BIO_write(next, &ctx->buf[ctx->buf_off], ctx->buf_len);
+    if (bytes_written <= 0) {
+      BIO_copy_next_retry(b);
+      break;
+    }
+    ctx->buf_off += bytes_written;
+    ctx->buf_len -= bytes_written;
+  }
+  return bytes_consumed;
+}
+
+static long enc_ctrl(BIO *b, int cmd, long num, void *ptr) {
+  GUARD_PTR(b);
+  long ret = 1;
+
+  BIO_ENC_CTX *ctx = BIO_get_data(b);
+  BIO *next = BIO_next(b);
+  if (ctx == NULL) {
+    return 0;
+  }
+
+  switch (cmd) {
+    case BIO_CTRL_RESET:
+      ctx->done = 0;
+      ctx->ok = 1;
+      ctx->buf_off = 0;
+      ctx->buf_len = 0;
+      OPENSSL_cleanse(ctx->buf, sizeof(ctx->buf));
+      if (!EVP_CipherInit_ex(ctx->cipher, NULL, NULL, NULL, NULL,
+                             EVP_CIPHER_CTX_encrypting(ctx->cipher))) {
+        return 0;
+      }
+      ret = BIO_ctrl(next, cmd, num, ptr);
+      break;
+    case BIO_CTRL_EOF:
+      if (ctx->done) {
+        ret = 1;
+      } else {
+        ret = BIO_ctrl(next, cmd, num, ptr);
+      }
+      break;
+    case BIO_CTRL_WPENDING:
+    case BIO_CTRL_PENDING:
+      // Return number of bytes left to process if we have anything buffered,
+      // else consult underlying BIO.
+      ret = ctx->buf_len;
+      if (ret <= 0) {
+        ret = BIO_ctrl(next, cmd, num, ptr);
+      }
+      break;
+    case BIO_CTRL_FLUSH:
+      ret = enc_flush(b, next, ctx);
+      if (ret <= 0) {
+        break;
+      }
+      // Flush the underlying BIO
+      ret = BIO_ctrl(next, cmd, num, ptr);
+      BIO_copy_next_retry(b);
+      break;
+    case BIO_C_GET_CIPHER_STATUS:
+      ret = (long)ctx->ok;
+      break;
+    // OpenSSL implements these, but because we don't need them and cipher BIO
+    // is internal, we can fail loudly if they're called. If this case is hit,
+    // it likely means you're making a change that will require implementing
+    // these.
+    case BIO_CTRL_DUP:
+    case BIO_CTRL_GET_CALLBACK:
+    case BIO_CTRL_SET_CALLBACK:
+    case BIO_C_DO_STATE_MACHINE:
+    case BIO_C_GET_CIPHER_CTX:
+      OPENSSL_PUT_ERROR(PKCS7, ERR_R_BIO_LIB);
+      return 0;
```

Comment:
NP: Do any of these cases potentially need to be propagated to `next` (as we do for default) instead of producing an error.

Reply:
i don't think so. i want to error out here so we are made aware of a need to implement these operations if something depending on us needs them.